Fix wall ordering

BagelOrb opened the pull request:
This PR builds on https://github.com/Ultimaker/CuraEngine/pull/1554

This PR solves a number of bugs all relating to the order in which walls are printed:
- Arachne splits wall polygons into several segments with extra starts/stops when Optimize Wall Printing Order is enabled, i.e. when Order Walls by Inset is disabled.
- Arachne then printed the separate segments belonging to each outline/hole region together, rather than printing whole polygons. It enforced ordering by region harder than stitching wall lines into polygons.
- The order in which walls are printed had some bugs in 4.13 and in Arachne which could cause some polygons to be printed out of order (inner to outer vs outer to inner)
- The gap filler walls were printed after the skin in 4.13, but in Arachne even before some of the walls when order was set to Outer to Inner.
- Arachne would sometimes leave tiny travel gaps in between a single wall 'gap filler', because stitching wasn't implemented properly.
- Wall lines were sometimes printed in the wrong direction (counter clockwise instead of clockwise).

Code changes:
- This PR applies the PolylineStitcher introduced in https://github.com/Ultimaker/CuraEngine/pull/1554 to form closed wall polygons before simplifying them en passing them to the optimizer.
- Because a stitched ExtrusionLine can have multiple region_id s associated with it, the InsetOrderOptimizer needed to be overhauled.
- I've implemented a proper weak order based on the adjacency between inset polygons, which is computed in InsetOrderOptimizer and used in PathOrderOptimizer.

Behavior changes:
- Gap fillers should always be printed after their enclosing wall, and depending on whether we use Center Last ordering, after all walls.
- Optimize wall order enabled / Order walls by inset disabled now strictly enforces the Inner-to-outer order per region, while wall polygons pertaining to two regions are not cut up, but printed before/after the other walls pertaining to both those regions.

Benefits:
- The quality of surfaces is better for complex layers with holes in them and/or with thin parts.
- The travel time is reduced.

Possibly affected behavior:
- All wall ordering, z seam config etc.
- of each wall feature type: Walls, infill walls, skin walls, support walls, support bottom walls, support infill walls, raft walls


Before:
![image](https://user-images.githubusercontent.com/8895761/150778321-233242ca-a52c-4ada-a8e9-d0480369be8d.png)

After:
![image](https://user-images.githubusercontent.com/8895761/150778369-9d029c3b-636c-4acf-8efe-5500a7dff46f.png)


Before:
![image](https://user-images.githubusercontent.com/8895761/150778643-6059d50e-8e43-4fb6-b867-219889dcd51c.png)

After:
![image](https://user-images.githubusercontent.com/8895761/150778747-bf62f895-c6be-4cf9-81be-ba09d81cbd04.png)

BagelOrb commented:
Fixes:
https://github.com/Ultimaker/CuraEngine/issues/1495

BagelOrb commented:
Results of tests on Applicatas dataset (assembled from models from the internet). Tested on S5 with Generic PLA at 0.1mm layer height.

Travel times:
Benchy:
4.13: 9s
Arachne beta 2: 11s
this PR: 10s

ACAM:
4.13: 2:04
Arachne beta 2: 1:18
this PR: 1:32

airhose splitter
4.13: 11s
Arachne beta 2: 11s
this PR: 9s

Alahamra
4.13: 2s
Arachne beta 2: 4s
this PR: 3s

David:
4.13: 0:53
Arachne beta 2: 1:03
this PR: 0:58


Decking fixture:
4.13: 26s
Arachne beta 2: 29s
this PR: 27s

Jet engine Combustion_and_HPT_Casing
4.13: 32s
Arachne beta 2: 39s
this PR: *crash!* Looking into it....

rburema commented:
Sort of the same story for this one, except in this case the build-server is down for some reason. Can you run the test manually & report back? Otherwise we can do that, but it might slow down the process a bit.

BagelOrb commented:
Here's the order on a complex model with a lot of internal holes: (Outer to Inner, Optimize Wall Ordering, Gap fillers always (perhaps immediately) after enclosing wall)

Outer and inner walls are printed intermittently, but the ordering is preserved everywhere:
![image](https://user-images.githubusercontent.com/8895761/151194106-6641abac-9827-426e-91e2-0156b50d8e7e.png)

![image](https://user-images.githubusercontent.com/8895761/151194170-dd96a8a9-b30b-466b-b139-a191511fe979.png)

![image](https://user-images.githubusercontent.com/8895761/151194248-12b93887-198b-4b1b-a519-0cf4ab4d5a3b.png)

Now we can print some of the small islands of inner walls:
![image](https://user-images.githubusercontent.com/8895761/151194569-074cdb87-6f80-49c4-99e9-1f92dfc3bda9.png)

Here wall 2 is printed on the right side, then it takes a gap filler of index 2 to the left side, and then continues to print inset 3 for both islands, before moving on to a gap filler of index 2 farther away:
![image](https://user-images.githubusercontent.com/8895761/151194635-6513193b-0338-4776-bd55-b4cfcc6c4af6.png)

![image](https://user-images.githubusercontent.com/8895761/151194735-ecc30d5f-97d0-474f-aada-94574b93cbb7.png)

Notice how the gap filler of index 3 is printed together with the pieces of wall with index 3:
![image](https://user-images.githubusercontent.com/8895761/151194816-f0af389e-3d9a-44e3-98ea-fc9d7332d94d.png)

BagelOrb commented:
With the above model 
[UMS5_Stand.3mf.renamed.zip](https://github.com/Ultimaker/CuraEngine/files/7943039/UMS5_Stand.3mf.renamed.zip)

These are the travel times:
Inside to Outside: 0:49
Outside to Inside: 0:49
Center Last: 0:42
Arachne beta 2, Center Last: 0:53
4.13, Inner before outer: 0:41

richfelker commented:
Is this PR in a state where it's suitable for testing? I'm the reporter of #1495 and would like to test if so.

BagelOrb commented:
Yes it is. If you are able to test from source then we would be very interested in your findings.

Before it will be merged some developers will review my code and some more tests will be performed, but there's always a chance for bugs to fall through, so I would be very happy if you could help our testing efforts.

richfelker commented:
OK, I can confirm at least for my test case in that report that the problem is fixed.

richfelker commented:
Unfortunately it segfaulted trying to slice a Benchy.

richfelker commented:
The crash occurs with `skin_monotonic=true` which I always use. Here's the backtrace (sorry probably lacks sufficient debug info I didn't build with): 
[curacrash.txt](https://github.com/Ultimaker/CuraEngine/files/8006486/curacrash.txt)

BagelOrb commented:
Could you please save your project as a 3mf file, put it in a zip and upload it here?

richfelker commented:
I don't use or have the Cura gui, just CuraEngine. I can probably make you a minimal reproducing testcase with that.

richfelker commented:
OK, all that's required to reproduce the issue is `skin_monotonic=true` and `top_bottom_pattern_0=zigzag` with the baseline `fdmprinter.def.json` and the standard Benchy STL.

BagelOrb commented:
Thanks. This is a problem on master. Fixed in https://github.com/Ultimaker/CuraEngine/commit/fa196dbdcc24ad7f94f19c2bb2cddc3927eb9a41

richfelker commented:
OK, good to know. It was not present in master but appeared when checking out your PR branch which is a ways behind master and apparently doesn't merge cleanly. I'll try the merge or wait for you to get it merged upstream.

BagelOrb commented:
The bug was in master since somehwere last week, so perhaps you hadn't pulled he bug in yet. I'll merge master into this branch soon.

richfelker commented:
OK, I can confirm it's working well now.

casperlamboo reviewed (commented):
As discussed part I would submit (the incomplete) review earlier. Expect part 2/2 of the review soon.

BagelOrb commented on src/utils/PolylineStitcher.h:
```diff
@@ -102,6 +103,10 @@ class PolylineStitcher
                             { // it was already moved to output
                                 return true; // keep looking for a connection
                             }
+                            if (!canReverse(nearby) && ((nearby.point_idx == 0) == go_in_reverse_direction))
```
Fixed in 0f6e26daa187a2feb534c732115f38209d8b9cfa

BagelOrb commented on src/WallToolPaths.cpp:
```diff
@@ -104,6 +105,37 @@ const VariableWidthPaths& WallToolPaths::generate()
+        if (wall_idx >= wall_lines.size())
+        {
+            wall_lines.resize(wall_idx + 1);
+        }
```
I can't find that code and github can't find the commit. Looks like you're looking at outdated commits. Idk.

BagelOrb commented on src/WallToolPaths.cpp:
```diff
@@ -134,6 +152,11 @@ const VariableWidthPaths& WallToolPaths::generate()
+        for (ExtrusionLine& line : wall_lines)
+        {
+            line.inset_idx = wall_idx;
+        }
```
This is outdated code. At that point in time the stitcher didn't yet ensure that the inset_idx survived, so I set it back afterwards.

However, this code is gone now.

BagelOrb commented on src/WallToolPaths.cpp:
```diff
@@ -415,4 +415,156 @@ void WallToolPaths::stitchContours(const VariableWidthPaths& input, const coord_
+                svg.writeArrow(all_polygons[i][0], all_polygons[child_idx][1]);
+            }
+        }
+    }
```
Fixed in eabfeae578a6747293cf933e97f9ff4fdadc8b05

BagelOrb commented on src/WallToolPaths.h:
```diff
@@ -83,6 +84,30 @@ class WallToolPaths
+     * Odd walls should always go after their enclosing wall polygons.
+     * 
+     * \param outer_to_inner Whether the wall polygons with a lower inset_idx should go before those with a higher one.
+     */
```
A weak order is mathematically defined as that the order between two elements is not given for each combination of two elements. All elements in the natural numbers are ordered, so it's called a strong order. Not all elements in the toolpaths are ordered, because no order requirement is given between two walls of the same inset.

The function you suggest is already implemented in that branch. You are looking at outdated code.

casperlamboo replied:
Is a weak ordering the same as a partially ordered set? https://en.wikipedia.org/wiki/Partially_ordered_set

BagelOrb commented on src/utils/polygon.cpp:
```diff
@@ -1576,6 +1577,112 @@ void Polygons::splitIntoPartsView_processPolyTreeNode(PartsView& partsView, Poly
+    for (int i = 0; i < 2; i++)
+    {
+        
+
```
outdated

BagelOrb commented on src/utils/polygon.cpp:
```diff
@@ -1576,6 +1577,112 @@ void Polygons::splitIntoPartsView_processPolyTreeNode(PartsView& partsView, Poly
+        {
+            auto it = node_to_index.find(child);
+            assert(it != node_to_index.end() && "Each PolyNode should be mapped to the corresponding Polygon index!");
+            ret[index].emplace_back(it->second);
```
I'm not sure. This exact line doesn't appear anymore in the latest branch. Please check if the new code is still unclear to you.

BagelOrb commented on src/PathOrderOptimizer.h:
```diff
@@ -411,6 +571,18 @@ class PathOrderOptimizer
+     * 
+     * Insertion sort can be wildly inefficient when polylines haven't been stitched.
+     */
+    bool selection_optimization;
```
This is oudated code. The whole InsertionSort algorithm has been removed.

BagelOrb commented on src/utils/polygon.cpp:
```diff
@@ -1576,6 +1577,112 @@ void Polygons::splitIntoPartsView_processPolyTreeNode(PartsView& partsView, Poly
+            const ClipperLib::PolyNode* node = queue.front();
+            queue.pop();
+            for (auto child : node->Childs)
+                queue.emplace(child);
```
Fixed in 4b0fa036516aec6dfcd99ced30f40ebce8e37e86

BagelOrb commented on src/utils/ExtrusionLine.h:
```diff
@@ -189,6 +198,16 @@ struct ExtrusionLine
+    {
+        Polygon ret;
+        
+        for (const ExtrusionJunction& j : junctions)
+            ret.add(j.p);
```
Fixed in 4b0fa036516aec6dfcd99ced30f40ebce8e37e86

BagelOrb commented on src/utils/polygon.cpp:
```diff
@@ -1576,6 +1577,112 @@ void Polygons::splitIntoPartsView_processPolyTreeNode(PartsView& partsView, Poly
+    for (auto [p, i] : start_loc_to_index)
+        std::cerr << p << " to index " << i << '\n';
+    std::cerr <<'\n';
+ 
```
Yes. Marked as such in 4b0fa036516aec6dfcd99ced30f40ebce8e37e86

BagelOrb commented on src/utils/polygon.cpp:
```diff
@@ -1576,6 +1577,112 @@ void Polygons::splitIntoPartsView_processPolyTreeNode(PartsView& partsView, Poly
+    {
+        std::cerr << "Couldn't find match for node with locations:\n";
+        for (auto p : node->Contour)
+            std::cerr << Point(p) << ", ";
```
Fixed in 4b0fa036516aec6dfcd99ced30f40ebce8e37e86

BagelOrb commented on src/utils/polygon.cpp:
```diff
@@ -1576,6 +1577,112 @@ void Polygons::splitIntoPartsView_processPolyTreeNode(PartsView& partsView, Poly
+    }
+    std::cerr << "Among registered locations: \n";
+    for (auto [p, i] : start_loc_to_index)
+        std::cerr << p << " to index " << i << '\n';
```
Fixed in 4b0fa036516aec6dfcd99ced30f40ebce8e37e86

BagelOrb commented on src/utils/polygon.cpp:
```diff
@@ -1576,6 +1577,112 @@ void Polygons::splitIntoPartsView_processPolyTreeNode(PartsView& partsView, Poly
+            }
+        }
+        for (const ClipperLib::PolyNode* node : unprocessed)
+            queue.emplace(node);
```
Fixed in 4b0fa036516aec6dfcd99ced30f40ebce8e37e86

BagelOrb commented on src/WallToolPaths.h:
```diff
@@ -83,6 +84,30 @@ class WallToolPaths
 
+    /*!
+     * Get the order constraints of the insets assuming the Wall Ordering is outer to inner.
+     * Each returned pair consists of adjacent wall lines where the left has an inset_idx one lower than the right.
```
There is no `include_transitive` anymore in the current branch.

Ghostkeeper commented:
Fixes https://github.com/Ultimaker/CuraEngine/issues/1552.

casperlamboo commented on src/InsetOrderOptimizer.cpp:
```diff
@@ -99,127 +107,234 @@ bool InsetOrderOptimizer::optimize(const WallType& wall_type)
-    return outer_region_id;
-}
+    };
+    
```
I would add a comment here that you use a grid to find nearby points. These nearby points are most-likely neighbour-toolpaths.

BagelOrb replied:
Done

casperlamboo commented on src/SkeletalTrapezoidation.h:
```diff
@@ -577,7 +577,7 @@ class SkeletalTrapezoidation
      * add a new toolpath segment, defined between two extrusion-juntions
      */
-    void addToolpathSegment(const ExtrusionJunction& from, const ExtrusionJunction& to, bool is_odd, bool force_new_path);
+    void addToolpathSegment(const ExtrusionJunction& from, const ExtrusionJunction& to, bool is_odd, bool force_new_path, bool from_is_3way, bool to_is_3way);
```
The intended use for `from_is_3way` and `to_is_3way` is not clear to me at first glace. Maybe at them to the `\param` comments?

BagelOrb replied:
Done

casperlamboo commented on src/utils/PolylineStitcher.h:
```diff
@@ -147,8 +160,15 @@ class PolylineStitcher
+
                 if (closest_is_closing_polygon)
                 {
+                    if (go_in_reverse_direction)
+                    { // re-reverse chain to retain original direction
+                        // NOTE: not sure if this code could ever be reached, since if a polygon can be closed that should be already possible in the forward direction
+                        chain.reverse();
+                    }
```
I missed this from last PR, but allow me to comment here what I should have commented on the polylines PR.

```cpp
            for (bool go_in_reverse_direction : { false, true })
            {
                if (go_in_reverse_direction)
                { // try extending chain in the other direction
                    chain.reverse();
                }
                
                ...
                
```
I don't think this reversing behaviour is very clear as this is dependent on the ordering of the reversal. If you would loop though the ordering like `go_in_reverse_direction : { true, false }` then this trick wouldn't work; the `true` case mutates the chain, and therefor must be done last. 

Why not use `emplace_back` or `emplace_front` conditionally on `go_in_reverse_direction`?

casperlamboo replied:
If using `emplace_back` / `emplace_front` add too much performance/complexity/overhead then at the very least add a comment that and why the true case must be handled last in the for loop.

BagelOrb replied:
> Why not use emplace_back or emplace_front conditionally on go_in_reverse_direction?

Because `emplace_front()` has a linear time complexity, rather than the constant time complexity of `emplace_back`.

Ok I've added comments.

casperlamboo commented on src/WallToolPaths.cpp:
```diff
@@ -148,25 +224,59 @@ void WallToolPaths::pushToolPaths(VariableWidthPaths& paths)
+        if (is_contour)
+        {
+#ifdef DEBUG
+            for (const ExtrusionLine& line : inset)
+                for (const ExtrusionJunction& j : line)
+                    assert(j.w == 0);
```
```suggestion
            for (const ExtrusionLine& line : inset)
            {
                for (const ExtrusionJunction& j : line)
                {
                    assert(j.w == 0);
                }
            }
````

BagelOrb replied:
Done

casperlamboo commented on src/WallToolPaths.cpp:
```diff
@@ -148,25 +224,59 @@ void WallToolPaths::pushToolPaths(VariableWidthPaths& paths)
+    inner_contour.clear();
+    for (const VariableWidthLines& inset : toolpaths)
+    {
+        if (inset.empty()) continue;
```
```suggestion
        if (inset.empty()) 
        {
            continue;
        }
````

BagelOrb replied:
Done.

casperlamboo commented on src/SkeletalTrapezoidation.cpp:
```diff
@@ -1861,24 +1859,37 @@ void SkeletalTrapezoidation::addToolpathSegment(const ExtrusionJunction& from, c
+    if (generated_toolpaths[inset_idx].empty()
+        || generated_toolpaths[inset_idx].back().is_odd != is_odd
+        || generated_toolpaths[inset_idx].back().junctions.back().perimeter_index != inset_idx // inset_idx should always be consistent
+        || ( ! is_odd && ! from_is_3way && ! to_is_3way && ( // if this is a normal even line ...
+            from.region_id != to.region_id // It should always have a consistent region_id
+            || generated_toolpaths[inset_idx].back().junctions.back().region_id != to.region_id
+            ))
```
This condition is getting pretty long and complicated.

Might I suggest splitting parts of this condition in a variable(s). This reduces the complexity of the whole condition and documentation can be encoded in the variable name.

BagelOrb replied:
Done

BagelOrb commented on src/FffGcodeWriter.cpp:
```diff
@@ -1405,7 +1409,10 @@ void FffGcodeWriter::addMeshLayerToGCode(const SliceDataStorage& storage, const
     }
 
-    processIroning(mesh, layer, mesh_config.ironing_config, gcode_layer);
+    if (extruder_nr == mesh.settings.get<ExtruderTrain&>((mesh.settings.get<size_t>("roofing_layer_count") > 0)? "roofing_extruder_nr" : "top_bottom_extruder_nr").extruder_nr)
```
Done

BagelOrb commented on src/InsetOrderOptimizer.cpp:
```diff
@@ -99,127 +107,313 @@ bool InsetOrderOptimizer::optimize(const WallType& wall_type)
+    
+    if (center_last)
+    {
+        for (const ExtrusionLine* line : walls_to_be_added)
+            if (line->is_odd)
+                for (const ExtrusionLine* other_line : walls_to_be_added)
+                    if ( ! other_line->is_odd)
+                        order.emplace(std::make_pair(other_line, line));
```
Done

BagelOrb commented on src/InsetOrderOptimizer.cpp:
```diff
@@ -99,127 +107,234 @@ bool InsetOrderOptimizer::optimize(const WallType& wall_type)
+    
+    constexpr Ratio flow = 1.0_r;
+    
+    bool added_something = false;
```
`added_something` was a variable which was already there in FffGcodeWriter for a long time. The naming ofthe variable here is to be consistent with the naming in FffGcodeWriter. I don't think it's good to change that variable everywhere in this PR.

BagelOrb commented on src/InsetOrderOptimizer.cpp:
```diff
@@ -99,127 +107,234 @@ bool InsetOrderOptimizer::optimize(const WallType& wall_type)
+    { // compute max_line_w
+        for (const ExtrusionLine* line : input)
         {
-            AABB& aabb = region_ids_to_bboxes[line.region_id]; // Empty AABBs are default initialized when region_ids are encountered for the first time.
-            for (const auto& junction : line.junctions)
+            for (const ExtrusionJunction& junction : *line)
             {
-                aabb.include(junction.p);
+                max_line_w = std::max(max_line_w, junction.w);
             }
         }
     }
```
Done

BagelOrb commented on src/utils/ExtrusionLine.cpp:
```diff
@@ -40,19 +44,15 @@ coord_t ExtrusionLine::getMinimalWidth() const
 void ExtrusionLine::simplify(const coord_t smallest_line_segment_squared, const coord_t allowed_error_distance_squared, const coord_t maximum_extrusion_area_deviation)
 {
-    if (junctions.size() <= 3)
+    if (junctions.size() <= 2 + is_closed)
```
Done

BagelOrb commented on src/utils/ExtrusionLine.cpp:
```diff
@@ -40,19 +44,15 @@ coord_t ExtrusionLine::getMinimalWidth() const
         return;
     }
 
+    // TODO: allow for the first point to be removed in case of simplifying closed Extrusionlines.
```
It's still a TODO. I didn't want to burn my fingers on changing any simplify code, because historically that function has been very bug-prone!

I think there is code duplication between ExtrusionLine::simplify and Polygons::simplify. Perhaps we should merge those two into one templated function. However, I didn't want to add al that complexity to this PR.

What does the Cura team think about this? Should I fix the simplify function in this PR, or would that be out of scope?

BagelOrb replied:
It should be noted that before this PR the wall lines would be connected in the PathOrderOptimizer, so the simplification never happened across ends of polylines. In that regard I didn't introduce this bug, but only reduced its impact, seeing as there are now less ends to the extrusionLines because they are stitched.

BagelOrb commented on src/utils/polygon.cpp:
```diff
@@ -11,6 +11,8 @@
 #include "ListPolyIt.h"
 
 #include "PolylineStitcher.h"
+#include "logoutput.h"
+#include "SVG.h"
```
Fixed

BagelOrb commented on src/utils/polygon.h:
```diff
@@ -81,6 +82,8 @@ class ConstPolygonRef
     : path(const_cast<ClipperLib::Path*>(&polygon))
     {}
 
+    ConstPolygonRef() =delete; // you cannot have a reference without an object!
```
I changed this everywhere in polygon.h

BagelOrb commented on src/utils/polygon.h:
```diff
@@ -410,6 +413,8 @@ class PolygonRef : public ConstPolygonRef
     : ConstPolygonRef(*other.path)
     {}
 
+    PolygonRef() =delete; // you cannot have a reference without an object!
```
Done.

BagelOrb commented on src/FffGcodeWriter.cpp:
```diff
@@ -2896,11 +2868,15 @@ bool FffGcodeWriter::addSupportRoofsToGCode(const SliceDataStorage& storage, Lay
     }
     if (! roof_paths.empty())
```
Indeed, it's so that you don't overlook the `!`.
Changed everywhere in PR.

BagelOrb commented on src/InsetOrderOptimizer.cpp:
```diff
@@ -85,7 +93,7 @@ bool InsetOrderOptimizer::optimize(const WallType& wall_type)
         }
         else if(extruder_nr == wall_x_extruder_nr)
```
That line of code was unchanged. I'm not sure why, but Ruben seems to adhere to the code standard that there shouldn't be a space after a keyword. That's why I asked for a document of the code conventions.

BagelOrb commented on src/InsetOrderOptimizer.cpp:
```diff
@@ -99,127 +107,234 @@ bool InsetOrderOptimizer::optimize(const WallType& wall_type)
-    const bool retract_before_outer_wall = mesh.settings.get<bool>("travel_retract_before_outer_wall");
-    const coord_t wall_0_wipe_dist = mesh.settings.get<coord_t>("wall_0_wipe_dist");
-    for(size_t inset = start_inset; inset != end_inset; inset += direction)
+    for(size_t inset_idx = start_inset; inset_idx != end_inset; inset_idx += direction)
```
Done

BagelOrb commented on src/InsetOrderOptimizer.cpp:
```diff
@@ -99,127 +107,234 @@ bool InsetOrderOptimizer::optimize(const WallType& wall_type)
+        const coord_t wipe_dist = is_outer_wall && ! is_gap_filler ? wall_0_wipe_dist : wall_x_wipe_dist;
+        const bool retract_before = is_outer_wall ? retract_before_outer_wall : false;
+        
+        const bool alternate_direction_modifier = alternate_walls && (path.vertices->inset_idx % 2 == layer_nr % 2);
+        const bool backwards = path.backwards != alternate_direction_modifier;
```
`^` is a bitwise operator, whereas `!=` is the xor for boolean values. Now for boolean values there is only one bit of interest, so it would work, but idk - it seems like the inappropriate operator. We also wouldn't use the bitwise `&` or the bitwise `|`.

I read up on it and it turns out we also shouldn't be using the bitwise assignments `&=` and `|=` for that matter.

BagelOrb replied:
I've applied your changes, bit with `!=` instead of `^`.

casperlamboo replied:
You could also take the sum of all these boolean values and do a `mod 2` on the result, doesn't increase readability though.

BagelOrb replied:
Lol

BagelOrb commented on src/InsetOrderOptimizer.cpp:
```diff
@@ -99,127 +107,234 @@ bool InsetOrderOptimizer::optimize(const WallType& wall_type)
+        const bool linked_path = p_start != p_end;
+        
+        
+        added_something = true;
```
Done

BagelOrb commented on src/TopSurface.cpp:
```diff
@@ -135,11 +137,13 @@ bool TopSurface::ironing(const SliceMeshStorage& mesh, const GCodePathConfig& li
     }
     if(!ironing_paths.empty())
     {
-        const BinJunctions binned_paths = InsetOrderOptimizer::variableWidthPathToBinJunctions(ironing_paths);
```
Why are the ironing lines binned here?

BagelOrb commented on src/TopSurface.cpp:
```diff
@@ -38,13 +39,14 @@ void TopSurface::setAreasFromMeshAndLayerNumber(SliceMeshStorage& mesh, size_t l
 }
 
-bool TopSurface::ironing(const SliceMeshStorage& mesh, const GCodePathConfig& line_config, LayerPlan& layer) const
+bool TopSurface::ironing(const SliceDataStorage& storage, const SliceMeshStorage& mesh, const GCodePathConfig& line_config, LayerPlan& layer, const FffGcodeWriter& gcode_writer) const
```
I didn't touch the order of ironing lines, only of the walls of ironing.
